### ISSUE TITLE
sample_weights not passed to loss from dataset

### ISSUE THREAD

**Issue by shkarupa-alex** [tensorflow/tensorflow#47968]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Google Colab
- TensorFlow version (use command below): v2.4.1-0-g85c8b2a817f 2.4.1
- Python version: 3
- Bazel version (if compiling from source): No
- GCC/Compiler version (if compiling from source): No
- CUDA/cuDNN version: No
- GPU model and memory: No

**Describe the current behavior**
I use tf.data.Dataset to feed data into model. Besides the features and labels I need to use weights for each sample (image pixel). I've tried to pass that weighs according to https://www.tensorflow.org/api_docs/python/tf/keras/Model?hl=en#fit but it didn't work.

**Describe the expected behavior**
Third value returned from Dataset should be passed to loss `sample_weights` argument.

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/153csG3Qv5GiXduEERHkG1Np3pMM958dX?usp=sharing

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported on tf 2.3,2.4 and nightly, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/1f96bd12950032edcb8c9a160cc25581/untitled571.ipynb).

**Comment by shkarupa-alex** [comment]
It seems i found an answer.
When passing a function as loss it wrapped with `LossFunctionWrapper` inside `LossesContainer`.  LossFunctionWrapper intercepts `sample_weights` in __call__ method and applies them on the result of original loss function call.

The final loss therefore should be computed correctly while original loss function returns result with a shape broadcastable to sample_weights.

Issue may be closed.
